Make GITHUB_TOKEN optional

pawroman opened the pull request:
This also makes the token possible to pass in from CLI args.

I got bitten by this a few times, and decided that issuing a warning instead of failing would be a better choice in terms of user experience.

Happy to tweak the behavior to if you think the new default is not an improvement over the old one.

pawroman commented on src/checker.rs:
```diff
@@ -419,7 +437,7 @@ mod test {
 
     #[tokio::test]
-    async fn test_exponetial_backoff() {
+    async fn test_exponential_backoff() {
```
Just a typo fix.

pawroman commented on README.md:
```diff
@@ -78,14 +79,21 @@ This comparison is made on a best-effort basis. Please create a PR to fix outdat
+### CLI return codes
+
+- `0` for success (all links checked successfully or excluded/skipped as configured)
+- `1` for any failures
```
This seems to be true for other errors bubbled up to `main()` (such as invalid value for `--method`)

mre replied:
Good point. We could use a different exit codes for pure link errors and config errors.
Proposing `1` for link errors and `2` for config errors.

pawroman replied:
Because of [how it currently works in Rust (return 1 for any Result::Err bubbled up to main)](https://github.com/rust-lang/rust/blob/a85e94927622665a9e9022de0d33a890a2e32d43/library/std/src/process.rs#L1844) and how the lychee code is structured, I propose we reverse that.

E.g. `1` for config errors and unexpected runtime errors, and `2` for link errors. Rationale: it's way simpler to bubble all "unexpected" errors to main using the `Result` enum and rely on Rust to exit the process with status code `1`. Where we can control the status code easily without much effort, e.g. link errors, we can return `2`.

This could be expanded to be `3` for specific config errors in the future, if need be, but it would require extra effort to classify errors and to normalize the `?` handling in `main()/run()`.

I will make the change to show what I mean; we can then discuss if it's OK with you.

mre commented on src/checker.rs:
```diff
@@ -389,7 +407,7 @@ mod test {
     fn get_checker(allow_insecure: bool, custom_headers: HeaderMap) -> Checker<'static> {
         let checker = Checker::try_new(
-            "DUMMY_GITHUB_TOKEN".to_string(),
+            None,
```
Much better. 😃

mre commented on src/main.rs:
```diff
@@ -86,8 +89,16 @@ async fn run(cfg: Config, inputs: Vec<String>) -> Result<i32> {
         None
     };
+
+    if cfg.github_token.is_none() {
```
Guess the default use-case would be non-Github links so I'd remove that warning as it might be confusing. How about we extend the status message from above?
Right now it is
```rust
Status::Error("GitHub token not specified".to_string()),
```
How about
```rust
Status::Error("GitHub token not specified. Use `--github-token` flag / `GITHUB_TOKEN` env var.".to_string()),
```

mre commented on src/options.rs:
```diff
@@ -133,6 +133,15 @@ pub(crate) struct Config {
     pub base_url: Option<String>,
+
+    #[structopt(
+        short,
```
Personally I'd remove the `short` option in case we'd like to support an option where `-g` is fitting. Can't think of anything right now but I'd rather keep the set of options small for now and make it quite explicit that we're passing a github token here.
I'm fine either way, though. 😉

pawroman replied:
Makes sense to me :+1:

mre commented:
Sorry, I merged in #23 so you might have to reintegrate `master`. 😅 
Added some comments. Let me know what you think. Otherwise PR looks good.

pawroman commented:
> Sorry, I merged in #23 so you might have to reintegrate `master`. sweat_smile
> Added some comments. Let me know what you think. Otherwise PR looks good.

Addressed the comments now and rebased against master.

mre commented on src/checker.rs:
```diff
@@ -445,7 +465,10 @@ mod test {
+
+        // on slow connections, this might take a bit longer than nominal backed-off timeout (7 secs)
+        assert!(end.as_secs() >= 7);
+        assert!(end.as_secs() <= 8);
```
Think there's a way to specify an epsilon. 🤔

mre replied:
Nope. There are [some workarounds](https://stackoverflow.com/questions/30856285/assert-eq-with-floating-point-numbers-and-delta), but nothing built-in. 😬

mre commented on src/main.rs:
```diff
@@ -20,6 +19,17 @@ use checker::{Checker, Excludes, Status};
+/// A C-like enum that can be cast to `i32` and used as process exit code.
+enum ExitCode {
+    Success = 0,
+    // NOTE: exit code 1 is used for any `Result::Err` bubbled up to `main()` using the `?` operator.
```
Thanks for documenting this behavior.

mre commented on src/main.rs:
```diff
@@ -123,7 +133,12 @@ async fn run(cfg: Config, inputs: Vec<String>) -> Result<i32> {
+
+    match success {
+        true => Ok(ExitCode::Success as i32),
+        false => Ok(ExitCode::LinkCheckFailure as i32),
```
Hey that's kinda cool!

mre commented on tests/cli.rs:
```diff
@@ -28,4 +27,73 @@ mod cli {
             .stdout(contains("Successful: 0"))
             .stdout(contains("Errors: 0"));
     }
+
```
These tests are all great!

mre commented:
Good work my sir. Chapeau. 🥂🧐🎩
